File: requirement_tests.R
library(RSelenium)
library(testthat)
library(askpass)
library(urltools)

mainpath_state <- TRUE
default_wait <- 2 #If a target system is running slow, bump this to increase the wait globally
dataverse_name <- "rselenium-test-dataverse"
dataset_name <- "rselenium-test-dataset"
dataset_id <- "" #set by code
template_id <- "" #set by code
root_dv_id <- "1" #assumed
login_user_api_token <- "" #set by code

source('functions/api_dataverse.R')
source('functions/test_functions_dataset.R')
source('functions/test_functions_dataverse.R')

#########################
### Requirement Tests ###
#########################

r01alt_mainpath_builtin_auth <- function() {
  username <- askpass("Username for builtin account")
  password <- askpass("Password for builtin account")
  #username <- ''
  #password <- ''
  
  sesh$navigate(paste(dv_server_url,'/loginpage.xhtml?redirectPage=%2Fdataverse.xhtml', sep=''))
  #
  sesh$findElement(value='//*[@id="loginForm:credentialsContainer:0:credValue"]')$sendKeysToElement(list(username))
  sesh$findElement(value='//*[@id="loginForm:credentialsContainer:1:sCredValue"]')$sendKeysToElement(list(password, key="enter"))
  #NOTE: For some reason this button click works with directly setting username/password, but NOT when using askpass.
  #      currently the code is just triggering an enter key after entering the password, but we may have the issue again
  #      ...
  #      I'm pretty sure this is an old issue due to using safari
  
  #sesh$findElement(using="xpath", value='//*[@id="loginForm:login"]')$clickElement()
  #sesh$screenshot(file = 'atest.png')
  
  sesh$findElement(value='//*[@id="dataverseDesc"]') #Find element to wait for load
  expect_identical(paste(dv_server_url,'/dataverse.xhtml', sep=''), toString(sesh$getCurrentUrl()))
}

r03_mainpath_create_sub_dataverse <- function() {
  sesh$navigate(dv_server_url)
  ### Main Landing Page ###
  sesh$findElement(value='//*[@id="addDataForm"]/div/button')$clickElement() #click add data
  sesh$findElement(value='//*[@id="addDataForm"]/div/ul/li[1]/a')$clickElement() #click new dataverse
  
  ### Create Dataverse Page ###
  sesh$findElement(value='//*[@id="dataverseForm:selectHostDataverse_input"]')$clearElement()
  sesh$findElement(value='//*[@id="dataverseForm:selectHostDataverse_input"]')$sendKeysToElement(list(paste(dv_props["host_dataverse"], sep='')))
  Sys.sleep(2) # wait for host list to load
  sesh$findElement(value='//*[@id="dataverseForm:selectHostDataverse_input"]')$sendKeysToElement(list(key = "enter"))
  Sys.sleep(.5) # wait after click for ui to be usable
  set_dataverse_metadata(add_string="create") #Metadata that is same for create/edit
 
  ### Test Save ###
  
  expect_identical(paste(sesh$findElement(value='//*[@id="messagePanel"]/div/div')$getElementAttribute("class")), "alert alert-success") #confirm success alert
  expect_identical(paste(dv_server_url,'/dataverse/',dataverse_name, '/', sep=''), toString(sesh$getCurrentUrl())) #confirm page
  
  test_dataverse_metadata(add_string="create")
  sesh$navigate(paste(dv_server_url, '/dataverse/', dataverse_name, sep=''))
  
  ### Dataverse Page - Publish ###
  
  sesh$findElement(value='//*[@id="actionButtonBlock"]/div/div/div[2]/button')$clickElement() #click publish
  sesh$findElement(value='//*[@id="dataverseForm:j_idt431"]')$clickElement() #confirm publish
  
  expect_identical(paste(sesh$findElement(value='//*[@id="messagePanel"]/div/div')$getElementAttribute("class")), "alert alert-success") #confirm success alert
}

r04_mainpath_edit_dataverse <- function() {
  sesh$navigate(paste(dv_server_url, '/dataverse/', dataverse_name, sep=''))
  Sys.sleep(default_wait)
  sesh$findElement(value='//*[@id="actionButtonBlock"]/div/div/div[2]/div[2]/button')$clickElement()
  sesh$findElement(value='//*[@id="dataverseForm:editInfo"]')$clickElement()
  
  set_dataverse_metadata()
  test_dataverse_metadata()
  
  Sys.sleep(1) #wait before switching pages in r05
}

r05_mainpath_create_metadata_template <- function() {
  sesh$navigate(dv_server_url)
  Sys.sleep(default_wait)

  sesh$findElement(value='//*[@id="actionButtonBlock"]/div/div[2]/div[2]/div/button')$clickElement() #click dataverse edit button
  sesh$findElement(value='//*[@id="dataverseForm:manageTemplates"]')$clickElement() #click manage templates
  Sys.sleep(default_wait)
  sesh$findElement(value='//*[@id="manageTemplatesForm"]/div[1]/div/a')$clickElement() #click create dataset template
  Sys.sleep(default_wait)
  sesh$findElement(value='//*[@id="templateForm:templateName"]')$sendKeysToElement(list("test template create")) #Create template title
  set_dataset_metadata_edit(add_string='create', xpath_dict=ds_template_xpaths)

  sesh$findElement(value='//*[@id="templateForm:j_idt892"]')$clickElement() #click "Save + Add Terms"

  expect_identical(paste(sesh$findElement(value='//*[@id="messagePanel"]/div/div[1]')$getElementAttribute("class")), "alert alert-success") #confirm success alert
  template_id <<- toString(param_get(toString(sesh$getCurrentUrl()), c("id")))
  # sesh$navigate(dv_server_url)
  # Sys.sleep(default_wait)
  # sesh$findElement(value='//*[@id="actionButtonBlock"]/div/div[2]/div[2]/div/button')$clickElement() #click dataverse edit button
  # sesh$findElement(value='//*[@id="dataverseForm:manageTemplates"]')$clickElement() #click manage templates
  # Sys.sleep(9999999999)
  #TODO: I'm not sure what ownerId refers to. We may need to go through the UI (like we probably will have to for delete anyways...)
  sesh$navigate(paste(dv_server_url, '/template.xhtml?id=', template_id, '&ownerId=1&editMode=METADATA', sep=''))
  test_dataset_metadata(add_string='create', xpath_dict=ds_template_xpaths)
}

r09_mainpath_create_dataset <- function() {
  sesh$navigate(paste(dv_server_url, '/dataverse/', dataverse_name, sep=''))
  
  sesh$findElement(value='//*[@id="addDataForm"]/div/button')$clickElement() #click add data
  sesh$findElement(value='//*[@id="addDataForm"]/div/ul/li[2]/a')$clickElement() #click new dataset
  
  set_dataset_metadata_create(add_string='create')
  
  expect_identical(paste(sesh$findElement(value='//*[@id="messagePanel"]/div/div[1]')$getElementAttribute("class")), "alert alert-success") #confirm success alert
  
  ### Get dataset id from permissions page url for later uses ###
  dataset_id <<- sub(".*=", "", sesh$findElement(value='//*[@id="datasetForm:manageDatasetPermissions"]')$getElementAttribute("href")) 
  # print(dataset_id)
  
  sesh$findElement(value='//*[@id="actionButtonBlock"]/div[1]/div/a')$clickElement() #click publish

  sesh$findElement(value='//*[@id="datasetForm:j_idt2547"]')$clickElement() #click publish confirm

  sesh$findElement(value='label-default', using='class name') #Find element to wait for load. May trigger prematurely with files added.
  expect_identical(toString(sesh$findElement(value='//*[@id="title-label-block"]/span')$getElementText()), "Version 1.0") #Test dataset published
  
  sesh$findElement(value='//*[@id="editDataSet"]')$clickElement() #click add data
  sesh$findElement(value='//*[@id="datasetForm:editMetadata"]')$clickElement() #click edit dataset
  
  test_dataset_metadata(add_string='create', is_update=FALSE, xpath_dict=ds_edit_xpaths)
  
  sesh$findElement(value='//*[@id="datasetForm:cancel"]')$clickElement() #click out after testing data
  
  Sys.sleep(default_wait)
}

r10_mainpath_edit_dataset <- function() {
  sesh$findElement(value='//*[@id="editDataSet"]')$clickElement() #click add data
  sesh$findElement(value='//*[@id="datasetForm:editMetadata"]')$clickElement() #click new dataset

  set_dataset_metadata_edit(add_string='edit', xpath_dict=ds_edit_xpaths)
  sesh$findElement(value='//*[@id="datasetForm:saveBottom"]')$clickElement() #click to create dataset

  sesh$findElement(value='//*[@id="actionButtonBlock"]/div[1]/div/a')$clickElement() #click publish

  sesh$findElement(value='//*[@id="datasetForm:j_idt2547"]')$clickElement() #click publish confirm
  
  sesh$findElement(value='label-default', using='class name') #Find element to wait for load. May trigger prematurely with files added.
  expect_identical(toString(sesh$findElement(value='//*[@id="title-label-block"]/span')$getElementText()), "Version 1.1") #Test dataset published
  
  sesh$findElement(value='//*[@id="editDataSet"]')$clickElement() #click add data
  sesh$findElement(value='//*[@id="datasetForm:editMetadata"]')$clickElement() #click new dataset
  
  test_dataset_metadata(add_string='edit', is_update=TRUE, xpath_dict=ds_edit_xpaths) 

  sesh$findElement(value='//*[@id="datasetForm:cancelTop"]')$clickElement() #click cancel out of edit after testing
}


#############################################
### Requirement Test Additional Functions ###
#############################################

# TODO:
# - Begin implementing R03-R25
# - How do I handle cleanup?
#   - Probably best to tie it into the call_mainpath error code? Try to do deletes if possible.
#   - I need to delete datasets AND a dataverse

load_dataverse_admin_info_from_file <- function() {
  #Currently this is programmed to read from a secret_table file located at a fixed directory
  #TODO: This needs to be made dynamic once I understand how to do this better in rstudio
  dframe <- read.table(file='/Users/madunlap/Documents/GitHub/secret_table.txt',header=FALSE,
                       sep='=',col.names=c('key','value'))
  dtable <- data.table(dframe,key='key')
  
  #TODO: remove api token if we end up not needing it (we pull the user's instead)
  dv_server_url <<- dtable["DATAVERSE_SERVER_URL"]$value
  dv_admin_api_token <<- dtable["DATAVERSE_API_TOKEN"]$value
}

begin_user_browser <- function() {
  sesh <<- remoteDriver(
    remoteServerAddr = "localhost",
    port = 4444L,
    browserName = "chrome"
  )
  #sesh$errorDetails()
  sesh$open(silent=TRUE)
  #NOTE: This sets the timeout used to search for elements. This does not apply to other things like checking the current URL.
  #      In some places in this code we check for an element before checking the URL to leverage this dynamic check instead of setting explicit waits
  sesh$setTimeout("implicit", milliseconds=20000) #Set very high to handle dataverse publish steps
  # implicit_timeouts(sesh, 15000)
  #sesh$getStatus()
}

# Wrapper for all mainpath test functions. Allows us to not rewrite the same code over-and-over for all tests
call_mainpath <- function(FUN) {
  if(!mainpath_state) { return(mainpath_state) } #Fails if previous mainpath tests have failed
  
  mainpath_state <<- tryCatch({
    FUN()
    return(TRUE)
  }, expectation_failure=function(e) { #catch testthat failures
    print(e)
    return(FALSE)
  },error = function(e) { #catch general errors
    print(e)
    print(sesh$errorDetails()$localizedMessage)
    return(FALSE)
  })
  
  return(mainpath_state)
}

clean_up_mainpath <- function(do_ds=TRUE, do_dv=TRUE, do_tmp=TRUE) {
  if(do_ds) {
    tryCatch({
      destroy_dataset(dataset_id, dv_server_url, login_user_api_token) #dv_admin_api_token)
    }, error = function(e) { #print error
      print("Dataset Destroy Error")
      print(e)
    })
  }
  if(do_dv) {
    tryCatch({
      delete_dataverse(dataverse_name, dv_server_url, login_user_api_token) #dv_admin_api_token)
      sesh$navigate(dv_server_url)
    }, error = function(e) { #print error
      print("Dataverse Delete Error")
      print(e)
    })
  }
  if(do_tmp) {
    delete_template_via_ui(id=template_id, dv_id=root_dv_id) #based on the assu
    # tryCatch({
    #   delete_dataset_template(template_id, dv_server_url, login_user_api_token) #dv_admin_api_token)
    #   sesh$navigate(dv_server_url)
    # }, error = function(e) { #print error
    #   print("Dataverse Delete Error")
    #   print(e)
    # })
  }
}

#We get the api token for the logged in user
get_api_token <- function() {
  sesh$navigate(paste(dv_server_url,'/dataverseuser.xhtml?selectTab=apiTokenTab', sep=''))
  #Note: This test assumes you have already clicked "Create Token" with this account.
  login_user_api_token <<- toString(sesh$findElement(value='//*[@id="apiToken"]/pre/code')$getElementText())
}

##This is not currently needed for our requirements. It was built to get around permissions issues for our admin off the root dataverse 
# test_delete_dataverse <- function() {
#   sesh$navigate(paste(dv_server_url,'/dataverse/',dataverse_name, sep=''))
#   
#   Sys.sleep(default_wait)
#   
#   sesh$findElement(value='//*[@id="actionButtonBlock"]/div/div/div[2]/div/button')$clickElement() #click edit
#   sesh$findElement(value='//*[@id="dataverseForm:deleteDataset"]')$clickElement() #click delete
#   sesh$findElement(value='//*[@id="dataverseForm:j_idt435"]')$clickElement() #click confirm
#   
#   Sys.sleep(default_wait)
#   
#   expect_identical(paste(dv_server_url,'/dataverse/root/', sep=''), toString(sesh$getCurrentUrl())) #confirm page
#   expect_identical(paste(sesh$findElement(value='//*[@id="messagePanel"]/div/div')$getElementAttribute("class")), "alert alert-success") #confirm success alert
#   
# }

File: reticulate/python/requirement_tests.py
import time, unittest, os, traceback

#TODO: These import paths will almost certainly break when running this code outside R. Not sure how to fix this.
#      ... Ideally there would be some way to tell reticulate what directory to work from?
import python.dataset_test_helper as dst
import python.dataverse_test_helper as dvt

from django.contrib.staticfiles.testing import StaticLiveServerTestCase
from django.test import override_settings
#from seleniumwire import webdriver
from selenium import webdriver
from selenium.common.exceptions import NoSuchElementException
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.ui import WebDriverWait, Select
from webdriver_manager.chrome import ChromeDriverManager

#TODO: 
# - How am I handling one test failing causing the others to not run?
#   - Should we build it into python or r?
# - Can we make this code useful for people testing without R?

exec(compile(source=open('/Users/madunlap/Documents/GitHub/secret_table_user_test.py').read(), filename='.', mode='exec')) #Reads file with secret info.

class RequirementTests:
  sesh = None
  test_case = None
  failed = False
  test_case = unittest.TestCase()
  username = None
  password = None
  
  #TODO: we may want to reference our dataset title, dataverse name and dataverse identifier from the dictionaries instead of a global like before
  # ...  or we completely avoid the dictionaries. it probably can't live in two places.
  
  def r01alt_mainpath_builtin_auth(self, tc):
    self.sesh.get(f'{DV_URL}/loginpage.xhtml?redirectPage=%2Fdataverse.xhtml')
    self.sesh.find_element('xpath', '//*[@id="loginForm:credentialsContainer:0:credValue"]').send_keys(self.username)
    self.sesh.find_element('xpath', '//*[@id="loginForm:credentialsContainer:1:sCredValue"]').send_keys(self.password)
    self.sesh.find_element('xpath', '//*[@id="loginForm:login"]').click()
    
    self.sesh.find_element('xpath', '//*[@id="dataverseDesc"]') #Find element to wait for load
    tc.assertEqual(f'{DV_URL}/dataverse.xhtml', self.sesh.current_url)
  
  ### UNTESTED REQUIREMENT TESTS  
  def r03_mainpath_create_sub_dataverse(self, tc):
    self.sesh.get(DV_URL)
    ### Main Landing Page ###
    self.sesh.find_element('xpath', '//*[@id="addDataForm"]/div/button').click() #click add data
    self.sesh.find_element('xpath', '//*[@id="addDataForm"]/div/ul/li[1]/a').click() #click new dataverse
    
    ### Create Dataverse Page ###
    self.sesh.find_element('xpath', '//*[@id="dataverseForm:selectHostDataverse_input"]').clear()
    self.sesh.find_element('xpath', '//*[@id="dataverseForm:selectHostDataverse_input"]').send_keys(dv_props["host_dataverse"])
    time.sleep(2) # wait for host list to load
    self.sesh.find_element('xpath', '//*[@id="dataverseForm:selectHostDataverse_input"]').send_keys(Keys.ENTER)
    time.sleep(.5) # wait after click for ui to be usable
    dvt.set_dataverse_metadata(self.sesh, tc, add_string="create") #Metadata that is same for create/edit
   
    ### Test Save ###
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="messagePanel"]/div/div').get_attribute("class"), "alert alert-success") #confirm success alert
    print("requirement_test dvt.dataverse_name:" + dvt.dataverse_name)
    tc.assertEqual(DV_URL+'/dataverse/'+dvt.dataverse_name+'/', self.sesh.current_url) #confirm page
    
    test_dataverse_metadata(self.sesh, tc, add_string="create")
    self.sesh.get(DV_URL+'/dataverse/'+dvt.dataverse_name)
    
    ### Dataverse Page - Publish ###
    
    self.sesh.find_element('xpath', '//*[@id="actionButtonBlock"]/div/div/div[2]/button').click() #click publish
    self.sesh.find_element('xpath', '//*[@id="dataverseForm:j_idt431"]').click() #confirm publish
    
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="messagePanel"]/div/div').get_attribute("class"), "alert alert-success") #confirm success alert
    
    dataverse_id <<- sub(".*=", "", self.sesh.find_element('xpath', '//*[@id="dataverseForm:themeWidgetsOpts"]').get_attribute("href")) 
  
  def r04_mainpath_edit_dataverse(self, tc):
    self.sesh.get(DV_URL+'/dataverse/'+dvt.dataverse_name)
    time.sleep(default_wait)
    self.sesh.find_element('xpath', '//*[@id="actionButtonBlock"]/div/div/div[2]/div[2]/button').click()
    self.sesh.find_element('xpath', '//*[@id="dataverseForm:editInfo"]').click()
    
    set_dataverse_metadata(self.sesh, tc)
    test_dataverse_metadata(self.sesh, tc)
    
    time.sleep(1) #wait before switching pages in r05
  
  def r05_mainpath_create_metadata_template(self, tc):
    self.sesh.get(DV_URL+'/dataverse/'+dvt.dataverse_name)
    time.sleep(default_wait)
  
    self.sesh.find_element('xpath', '//*[@id="actionButtonBlock"]/div/div/div[2]/div[2]/button').click() #click dataverse edit button
    self.sesh.find_element('xpath', '//*[@id="dataverseForm:manageTemplates"]').click() #click manage templates
    time.sleep(default_wait)
    self.sesh.find_element('xpath', '//*[@id="manageTemplatesForm"]/div[1]/div/a').click() #click create dataset template
    time.sleep(default_wait)
    self.sesh.find_element('xpath', '//*[@id="templateForm:templateName"]').send_keys("test template create") #Create template title
    set_dataset_metadata_edit(self.sesh, tc, add_string='create', xpath_dict=ds_template_xpaths)
    
    self.sesh.find_element('xpath', '//*[@id="templateForm:j_idt620:0:j_idt623:0:instr"]').click()
    time.sleep(.2)
  #TODO: What?
    self.sesh.switch_to.active_element.send_keys("test"+"WOWOWOWOW")
    time.sleep(999999999999)
    
      
    self.sesh.find_element('xpath', '//*[@id="templateForm:j_idt892"]').click() #click "Save + Add Terms"
  
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="messagePanel"]/div/div[1]').get_attribute("class"), "alert alert-success") #confirm success alert
    
  #TODO: Replace param_get with a new way to get the query params from a url
    template_id <<- param_get(self.sesh.current_url, "id")
  
    set_template_license(self.sesh, tc, add_string='create')
    self.sesh.find_element('xpath', '//*[@id="templateForm:j_idt893"]').click() #click "Save Dataset Template" (which actually just saves the license)
    
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="messagePanel"]/div/div[1]').get_attribute("class"), "alert alert-success") #confirm success alert
    time.sleep(default_wait) #not sure why I have to do this but ok? Shouldn't confirming the alert work?
    
    self.sesh.get(DV_URL+'/template.xhtml?id='+template_id+'&ownerId='+dataverse_id+'&editMode=METADATA')
    test_dataset_metadata(self.sesh, tc, add_string='create', xpath_dict=ds_template_xpaths)
    
    self.sesh.get(DV_URL+'/template.xhtml?id='+template_id+'&ownerId='+dataverse_id+'&editMode=LICENSE')
    test_template_license(self.sesh, tc, add_string='create')
  
  def r06_mainpath_edit_metadata_template(self, tc):
    self.sesh.get(DV_URL+'/template.xhtml?id='+template_id+'&ownerId='+dataverse_id+'&editMode=METADATA')
    set_dataset_metadata_edit(self.sesh, tc, add_string='edit', xpath_dict=ds_template_xpaths)
    self.sesh.find_element('xpath', '//*[@id="templateForm:j_idt893"]').click() #click "Save + Add Terms"
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="messagePanel"]/div/div[1]').get_attribute("class"), "alert alert-success") #confirm success alert
    
    self.sesh.get(DV_URL+'/template.xhtml?id='+template_id+'&ownerId='+dataverse_id+'&editMode=LICENSE')
    set_template_license(self.sesh, tc, add_string='edit')
    self.sesh.find_element('xpath', '//*[@id="templateForm:j_idt893"]').click() #click "Save Changes"
    time.sleep(default_wait) #not sure why I have to do this but ok? Shouldn't confirming the alert work?
    
    self.sesh.get(DV_URL+'/template.xhtml?id='+template_id+'&ownerId='+dataverse_id+'&editMode=METADATA')
    test_dataset_metadata(self.sesh, tc, add_string='edit', xpath_dict=ds_template_xpaths)
    
    self.sesh.get(DV_URL+'/template.xhtml?id='+template_id+'&ownerId='+dataverse_id+'&editMode=LICENSE')
    test_template_license(self.sesh, tc, add_string='edit')
  
  def r09_mainpath_create_dataset(self, tc):
    self.sesh.get(DV_URL+'/dataverse/'+dvt.dataverse_name)
    
    self.sesh.find_element('xpath', '//*[@id="addDataForm"]/div/button').click() #click add data
    self.sesh.find_element('xpath', '//*[@id="addDataForm"]/div/ul/li[2]/a').click() #click new dataset
    
    set_dataset_metadata_create(self.sesh, tc, add_string='create')
    
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="messagePanel"]/div/div[1]').get_attribute("class"), "alert alert-success") #confirm success alert
    
    ### Get dataset id from permissions page url for later uses ###
    dataset_id <<- sub(".*=", "", self.sesh.find_element('xpath', '//*[@id="datasetForm:manageDatasetPermissions"]').get_attribute("href")) 
    # print(dataset_id)
    
    self.sesh.find_element('xpath', '//*[@id="actionButtonBlock"]/div[1]/div/a').click() #click publish
  
    self.sesh.find_element('xpath', '//*[@id="datasetForm:j_idt2547"]').click() #click publish confirm
  
    self.sesh.find_element('xpath', 'label-default', using='class name') #Find element to wait for load. May trigger prematurely with files added.
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="title-label-block"]/span').text, "Version 1.0") #Test dataset published
    
    self.sesh.find_element('xpath', '//*[@id="editDataSet"]').click() #click add data
    self.sesh.find_element('xpath', '//*[@id="datasetForm:editMetadata"]').click() #click edit dataset
    
    test_dataset_metadata(self.sesh, tc, add_string='create', is_update=FALSE, xpath_dict=ds_edit_xpaths)
    
    self.sesh.find_element('xpath', '//*[@id="datasetForm:cancel"]').click() #click out after testing data
    
    time.sleep(default_wait)
  
  def r10_mainpath_edit_dataset(self, tc):
    self.sesh.find_element('xpath', '//*[@id="editDataSet"]').click() #click add data
    self.sesh.find_element('xpath', '//*[@id="datasetForm:editMetadata"]').click() #click new dataset
  
    set_dataset_metadata_edit(self.sesh, tc, add_string='edit', xpath_dict=ds_edit_xpaths)
    self.sesh.find_element('xpath', '//*[@id="datasetForm:saveBottom"]').click() #click to create dataset
  
    self.sesh.find_element('xpath', '//*[@id="actionButtonBlock"]/div[1]/div/a').click() #click publish
  
    self.sesh.find_element('xpath', '//*[@id="datasetForm:j_idt2547"]').click() #click publish confirm
    
    self.sesh.find_element('xpath', 'label-default', using='class name') #Find element to wait for load. May trigger prematurely with files added.
    tc.assertEqual(self.sesh.find_element('xpath', '//*[@id="title-label-block"]/span').text, "Version 1.1") #Test dataset published
    
    self.sesh.find_element('xpath', '//*[@id="editDataSet"]').click() #click add data
    self.sesh.find_element('xpath', '//*[@id="datasetForm:editMetadata"]').click() #click new dataset
    
    test_dataset_metadata(self.sesh, tc, add_string='edit', is_update=TRUE, xpath_dict=ds_edit_xpaths) 
  
    self.sesh.find_element('xpath', '//*[@id="datasetForm:cancelTop"]').click() #click cancel out of edit after testing
  
  #############################################
  ### Requirement Test Additional Functions ###
  #############################################
  
  def begin_user_browser(self, tc):
    options = Options()
    self.sesh = webdriver.Chrome(ChromeDriverManager().install(), options=options)
    self.sesh.implicitly_wait(20) #Set high for dataset publish. Maybe we should keep it low normally and set it up when needed?

  def test_caller(self, req_funct):
    if self.failed:
      return False
    
    try:
      req_funct(self.test_case)
      #test_assert(self.test_case)
      return True
    except Exception as e:
      self.failed = True
      #print(type(e).__name__ + ":", e)
      #print(traceback.format_exc())
      traceback.print_exc()
      return False
  
  #For use when getting the results of the individual steps doesn't matter. Not used by our R code  
  #TODO: make username/password optional somehow whe we do oauth
  def linear_test_runner(self, username, password):
    print(f"Set username/password: {self.set_username_password(username, password)}")
    print(f"Begin Browser: {self.test_caller(self.begin_user_browser)}")
    print(f"R01alt mainpath auth: {self.test_caller(self.r01alt_mainpath_builtin_auth)}")
    self.sesh.close()
    return self.failed
    
  def set_username_password(self, username, password):
    self.username = username
    self.password = password
    
    
    
    
    
    
